Chado buddy documentation (#90)
tripal/tripal_doc: docs/dev_guide/biodata/buddies/inject.rst

Injecting the Buddy Service
=============================

A typical use case for a Chado Buddy is in an importer.
Best practice is to inject the buddy service into your importer class.
Let's see how that is done.

First you will need to include these classes in your importer:

.. code::

>   use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
>   use Symfony\Component\DependencyInjection\ContainerInterface;
>   use Drupal\tripal_chado\ChadoBuddy\PluginManagers\ChadoBuddyPluginManager;

Reviewer: Include the actual buddies here as well to enable full type hinting and more clearly indicate dependencies.
```suggestion
  use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
  use Symfony\Component\DependencyInjection\ContainerInterface;
  use Drupal\tripal_chado\ChadoBuddy\PluginManagers\ChadoBuddyPluginManager;
  use Drupal\tripal_chado\Plugin\ChadoBuddy\‎ChadoDbxrefBuddy;
  use Drupal\tripal_chado\Plugin\ChadoBuddy\‎ChadoCvtermBuddy;
  use Drupal\tripal_chado\Plugin\ChadoBuddy\‎ChadoPropertyBuddy;
```

Then you will need to implement ``ContainerFactoryPluginInterface`` in your class. For example:

.. code::

>   abstract class MyImporter extends ChadoImporterBase implements ContainerFactoryPluginInterface {

Reviewer: Don't suggest abstract in the example since it is more likely they will want a fully fledged class here.

```suggestion
  class MyImporter extends ChadoImporterBase implements ContainerFactoryPluginInterface {
```

Next add a class variable to store the buddy service, and whichever buddy instance or instances you will need.

.. code::

>   /**
>    * Used to store the buddy service manager
>    */
>   protected object $buddy_manager;
> 
>   /**
>    * Used to store the buddy dbxref instance
>    */
>   protected object $dbxref_instance;
> 
>   /**
>    * Used to store the buddy cvterm instance
>    */
>   protected object $cvterm_instance;
> 
>   /**
>    * Used to store the buddy property instance
>    */
>   protected object $property_instance;

Reviewer: Update to match coding standards.
```suggestion
  /**
   * The Chado Buddy service manager
   *
   * @var Drupal\tripal_chado\ChadoBuddy\PluginManagers\ChadoBuddyPluginManager
   */
  protected ChadoBuddyPluginManager $buddy_manager;

  /**
   * An instance of the dbxref Chado Buddy.
   *
   * @var Drupal\tripal_chado\Plugin\ChadoBuddy\‎ChadoDbxrefBuddy
   */
  protected ‎ChadoDbxrefBuddy $dbxref_instance;

  /**
   * An instance of the cvterm Chado Buddy.
   *
   * @var Drupal\tripal_chado\Plugin\ChadoBuddy\‎ChadoCvtermBuddy
   */
  protected ‎ChadoCvtermBuddy $cvterm_instance;

  /**
   * An instance of the property Chado Buddy.
   *
   * @var Drupal\tripal_chado\Plugin\ChadoBuddy\‎ChadoPropertyBuddy
   */
  protected ‎ChadoPropertyBuddy $property_instance;
```
